mirgecom/steppers.py

```diff
@@ -64,17 +61,26 @@ def get_rhs():
     return get_rhs()
 
 
-def _force_evaluation(actx, state):
+def _strict_array_equal(actx, ary1, ary2):
+    """
+    Check if two arrays are equal.
+
+    If *ary1* and *ary2* are :mod:`pytato` arrays, additionally checks whether the
+    contained expressions are the same (not just whether they evaluate to the same
+    result).
+    """
     if actx is None:
-        return state
-    return thaw(freeze(state, actx), actx)
+        return np.array_equal(ary1, ary2)
+
+    return actx.to_numpy(actx.np.array_equal(ary1, ary2))[()]
```

Comment:
Pytato arrays use `==` for structural equality. This is one intentional departure from numpy. The ability to use them in mappings in mappings without gymnastics was the overriding concern.